### PR TITLE
onlykey-agent 1.1.14 (new formula)

### PR DESCRIPTION
**Pull request by activescott**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/onlykey-agent.rb
Comment (SMillerDev) [inline Formula/onlykey-agent.rb:11]: Why not the modern version?

File: Formula/onlykey-agent.rb
Comment (SMillerDev) [inline Formula/onlykey-agent.rb:15]:
```suggestion
  depends_on "openssl@3.0"
```

File: Formula/onlykey-agent.rb
Comment (SMillerDev) [inline Formula/onlykey-agent.rb:17]: Does it work with 3.10?

Comment (onlykey) [reply Formula/onlykey-agent.rb:17]: Yes

File: Formula/onlykey-agent.rb
Comment (onlykey) [reply Formula/onlykey-agent.rb:11]: onlykey-agent 1.1.14 works with gnupg2.3 and gnupg2.2 (LTS version)

**Comment by onlykey** [comment]
@activescott Thanks! we do want to do a brew formula. @SMillerDev Thanks for the review and recommended changes.

File: Formula/onlykey-agent.rb
Comment (SMillerDev) [reply Formula/onlykey-agent.rb:11]: Then it should use the newer version

File: Formula/onlykey-agent.rb
Comment (SMillerDev) [reply Formula/onlykey-agent.rb:15]: Yeah, it says that in the docs. That's why it should use that.

File: Formula/onlykey-agent.rb
Comment (SMillerDev) [reply Formula/onlykey-agent.rb:17]: Then it should use the newer version

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb:209]: @SMillerDev Thanks for all of your suggestions. On this one though, I was specifically using `` instead of `shell_output` since `shell_output` asserts the exit code is zero. In the case of this software, if the device happens to not be plugged in and unlocked at the time of install it returns a non-zero response code. Hence the regex below this line that looks for two different possible outputs.

**Comment by activescott** [comment]
Thanks for the feedback @onlykey and @SMillerDev. I'll look at getting these changes made and a rebuild working here tonight or tomorrow.

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb:17]: Using `openssl@3` caused conflicts:

```
brew audit --new-formula onlykey-agent && brew audit --strict --online onlykey-agent
onlykey-agent:
  * onlykey-agent contains conflicting version recursive dependencies:
      openssl@1.1, openssl@3
    View these with `brew deps --tree onlykey-agent`.
Error: 1 problem in 1 formula detected

```

```
brew deps --tree onlykey-agent
onlykey-agent
├── gnupg
│   ├── gettext
│   ├── gnutls
│   │   ├── ca-certificates
│   │   ├── gmp
│   │   ├── guile
│   │   │   ├── bdw-gc
│   │   │   ├── gmp
│   │   │   ├── libffi
│   │   │   ├── libtool
│   │   │   │   └── m4
│   │   │   ├── libunistring
│   │   │   ├── pkg-config
│   │   │   └── readline
│   │   ├── libidn2
│   │   │   ├── gettext
│   │   │   └── libunistring
│   │   ├── libtasn1
│   │   ├── libunistring
│   │   ├── nettle
│   │   │   └── gmp
│   │   ├── p11-kit
│   │   │   ├── ca-certificates
│   │   │   ├── libffi
│   │   │   └── libtasn1
│   │   └── unbound
│   │       ├── libevent
│   │       │   └── openssl@1.1
│   │       │       └── ca-certificates
│   │       ├── libnghttp2
│   │       └── openssl@1.1
│   │           └── ca-certificates
│   ├── libassuan
│   │   └── libgpg-error
│   ├── libgcrypt
│   │   └── libgpg-error
│   ├── libgpg-error
│   ├── libksba
│   │   └── libgpg-error
│   ├── libusb
│   ├── npth
│   └── pinentry
│       ├── libassuan
│       │   └── libgpg-error
│       └── libgpg-error
├── libusb
├── openssl@3
│   └── ca-certificates
├── python@3.10
│   ├── gdbm
│   ├── mpdecimal
│   ├── openssl@1.1
│   │   └── ca-certificates
│   ├── readline
│   ├── sqlite
│   │   └── readline
│   └── xz
├── rust
│   ├── libssh2
│   │   └── openssl@1.1
│   │       └── ca-certificates
│   ├── openssl@1.1
│   │   └── ca-certificates
│   └── pkg-config
└── six

```

File: Formula/onlykey-agent.rb
Comment (SMillerDev) [reply Formula/onlykey-agent.rb:209]: Since CI is the foremost user of `brew test` you can just assume that there is no device.

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb]: This was addressed in 271acd112276cc046f1328fb0ba367b7d446896f by using bare `gnupg` per a brew audit recommendation.

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb]: Addressed in 271acd112276cc046f1328fb0ba367b7d446896f

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb]: addressed in 271acd112276cc046f1328fb0ba367b7d446896f

**Comment by activescott** [comment]
I'm investigating the build failure on linux. At a glance looks like I need to declare a dependency for libhidapi-dev. I'll post back here when I've got something...

**Comment by activescott** [comment]
hopefully this latest commit (dad04f6cbbeb5f6fb9c6bb6415a38ef9a7a92603) fixes the linux build. I setup a container locally and it seemed to resolve it

**Comment by activescott** [comment]
@SMillerDev @onlykey the package is now building on mac and linux for me and CI so hopefully we're in good shape. let me know if you see anything else that needs adjusted.

File: Formula/onlykey-agent.rb
Comment (bayandin) [reply Formula/onlykey-agent.rb:79]: Or even `depends_on "libcython"`

File: Formula/onlykey-agent.rb
Comment (carlocab) [inline Formula/onlykey-agent.rb:24]:
```suggestion
  # https://docs.crp.to/onlykey-agent.html#installation
  depends_on "rust" => :build # for cryptography
  depends_on "gnupg"
  depends_on "libusb"
  depends_on "openssl@1.1" # for cryptography
  depends_on "python@3.10"
  depends_on "six"
```

File: Formula/onlykey-agent.rb
Comment (carlocab) [inline Formula/onlykey-agent.rb:34]: I think you need a `depends_on "hidapi"` for this to be useful?

File: Formula/onlykey-agent.rb
Comment (carlocab) [inline Formula/onlykey-agent.rb:84]: Maybe we can use the `docutils` formula?

File: Formula/onlykey-agent.rb
Comment (carlocab) [reply Formula/onlykey-agent.rb:199]: I think so. Though I thought `brew` excludes `wheel` automatically.

File: Formula/onlykey-agent.rb
Comment (carlocab) [inline Formula/onlykey-agent.rb:203]:
```suggestion
    ENV.append_to_cflags "-I#{Formula["libusb"].include}/libusb-1.0"
```

File: Formula/onlykey-agent.rb
Comment (carlocab) [inline Formula/onlykey-agent.rb:211]:
```suggestion
    assert_match("try unplugging and replugging your device", output)
```
You don't really actually need a regex here.

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb:199]: FYI: This came from `brew`, but I can try removing it and see if it builds.

Comment (activescott) [reply Formula/onlykey-agent.rb]: Resolved with 76928602fc5f47438142498b0e7217dac4a7b5a6

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb:34]: I'm happy to add it, but I don't know if this is necessary or not. FWIW I noticed it isn't in https://github.com/Homebrew/homebrew-core/blob/master/Formula/trezor-agent.rb which I know onlykey-agent is based on. @onlykey do you have advice here?

File: Formula/onlykey-agent.rb
Comment (cho-m) [reply Formula/onlykey-agent.rb]: We probably need to decide if we can/want to automatically exclude `wheel` in `brew`, otherwise the resource will end up returning on next version bump.

https://github.com/Homebrew/brew/blob/3.5.4/Library/Homebrew/utils/pypi.rb#L189
```rb
    exclude_packages += %W[#{main_package.name} argparse pip setuptools wsgiref].map { |p| Package.new p }
```

Comment (dtrodrigues) [reply Formula/onlykey-agent.rb]: `wheel` used to be excluded in that line, but it was decided in https://github.com/Homebrew/brew/pull/10849 to remove that logic.

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb:79]: I tried replacing it with `depends_on "libcython"` and brew test failed with the following on the ubuntu after a full build with the following:

```
#13 4.064   Minitest::Assertion: Expected /try unplugging and replugging your device/ to match "Traceback (most recent call last):
  File \"/home/linuxbrew/.linuxbrew/Cellar/onlykey-agent/1.1.14/bin/onlykey-agent\", line 33, in <module>
    sys.exit(load_entry_point('onlykey-agent==1.1.14', 'console_scripts', 'onlykey-agent')())
  File \"/home/linuxbrew/.linuxbrew/Cellar/onlykey-agent/1.1.14/bin/onlykey-agent\", line 25, in importlib_load_entry_point
    return next(matches).load()
  File \"/home/linuxbrew/.linuxbrew/opt/python@3.10/lib/python3.10/importlib/metadata/__init__.py\", line 171, in load
    module = import_module(match.group('module'))
  File \"/home/linuxbrew/.linuxbrew/opt/python@3.10/lib/python3.10/importlib/__init__.py\", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File \"<frozen importlib._bootstrap>\", line 1050, in _gcd_import
  File \"<frozen importlib._bootstrap>\", line 1027, in _find_and_load
  File \"<frozen importlib._bootstrap>\", line 1006, in _find_and_load_unlocked
  File \"<frozen importlib._bootstrap>\", line 688, in _load_unlocked
  File \"<frozen importlib._bootstrap_external>\", line 883, in exec_module
  File \"<frozen importlib._bootstrap>\", line 241, in _call_with_frames_removed
  File \"/home/linuxbrew/.linuxbrew/Cellar/onlykey-agent/1.1.14/libexec/bin/onlykey_agent.py\", line 1, in <module>
    import libagent.gpg
  File \"/home/linuxbrew/.linuxbrew/Cellar/onlykey-agent/1.1.14/libexec/lib/python3.10/site-packages/libagent/gpg/__init__.py\", line 22, in <module>
    import pkg_resources
  File \"/home/linuxbrew/.linuxbrew/Cellar/onlykey-agent/1.1.14/libexec/lib/python3.10/site-packages/pkg_resources/__init__.py\", line 3260, in <module>
    def _initialize_master_working_set():
  File \"/home/linuxbrew/.linuxbrew/Cellar/onlykey-agent/1.1.14/libexec/lib/python3.10/site-packages/pkg_resources/__init__.py\", line 3234, in _call_aside
    f(*args, **kwargs)
  File \"/home/linuxbrew/.linuxbrew/Cellar/onlykey-agent/1.1.14/libexec/lib/python3.10/site-packages/pkg_resources/__init__.py\", line 3272, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File \"/home/linuxbrew/.linuxbrew/Cellar/onlykey-agent/1.1.14/libexec/lib/python3.10/site-packages/pkg_resources/__init__.py\", line 581, in _build_master
    ws.require(__requires__)
  File \"/home/linuxbrew/.linuxbrew/Cellar/onlykey-agent/1.1.14/libexec/lib/python3.10/site-packages/pkg_resources/__init__.py\", line 909, in require
    needed = self.resolve(parse_requirements(requirements))
  File \"/home/linuxbrew/.linuxbrew/Cellar/onlykey-agent/1.1.14/libexec/lib/python3.10/site-packages/pkg_resources/__init__.py\", line 795, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'Cython>=0.23.4' distribution was not found and is required by onlykey
".
```

I'll try replacing it with `depends_on "cython"` instead...

Comment (activescott) [reply Formula/onlykey-agent.rb:79]: Looks like `depends_on "cython"` fails since that formula isn't updated with python 3.10 which I was [advised to update this package to in a prior comment on this PR](https://github.com/Homebrew/homebrew-core/pull/104096#discussion_r902676853). So some options to move forward:

1. Let `resource "Cython"` that `brew create` added stand 
2. Revert this package's dependency back to python@3.9 that `brew create` used
3. Attempt to update the cython formula to depend on pthon@3.10 and merge that before merging this one. I found the 3.9 bump at f23cda5147ab938d7977e34884d9535845b5353e and can make a similar change for 3.10 but and submit that - but admittedly I am not clear on the consequences of doing that.

Advice?

Comment (cho-m) [reply Formula/onlykey-agent.rb:79]: `cython` formula shouldn't be used unless it needs the executables (which is rare).

`libcython` formula (and probably `cython`) will probably fail due to them being keg-only, so the libraries cannot be found without some extra work. One option is to add a `.pth` file to virtualenv site-packages to pick up keg-only libraries.

Otherwise, we would need to use resource.

File: Formula/onlykey-agent.rb
Comment (cho-m) [reply Formula/onlykey-agent.rb:26]: PyPI (cython-)`hidapi` vendors a copy of `hidapi`.

It takes some extra work to override with system copy based on my previous PR for `liquidctl`. This is due to the extra command line arg (`--with-system-hidapi`) needed to force this:
https://github.com/Homebrew/homebrew-core/blob/693b3239cda90f7dbf0f42cd470333bb96a370e3/Formula/liquidctl.rb#L55-L60

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb:71]: Thank you @cho-m 🙇‍♂️ I'll move forward with resources for now. If maintainers have a concern please reply and I'd appreciate a bit more direction on the .pth/virtualenv site-packages suggestion if that is necessary as I'm not familiar with that.

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb:26]: Personally I prefer letting [`cython-hidapi`](https://github.com/trezor/cython-hidapi/blob/master/setup.py) vendor hidapi over monkey patching it, but maybe I'm missing something?

If using the homebrew hidapi is important enough maybe coming up with a patch to cython-hidapi to allow passing in the system hidapi include prefix better?

Comment (carlocab) [reply Formula/onlykey-agent.rb:26]: > Personally I prefer letting [`cython-hidapi`](https://github.com/trezor/cython-hidapi/blob/master/setup.py) vendor hidapi over monkey patching it, but maybe I'm missing something?

See https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae.

> If using the homebrew hidapi is important enough maybe coming up with a patch to cython-hidapi to allow passing in the system hidapi include prefix better?

If we can upstream it, that would be nice.

File: Formula/onlykey-agent.rb
Comment (carlocab) [reply Formula/onlykey-agent.rb:71]: > One option is to add a `.pth` file to virtualenv site-packages to pick up keg-only libraries.

I think `brew` [handles something like this](https://github.com/Homebrew/brew/blob/976681134384abf0466bbf3ce4acdb880a262d84/Library/Homebrew/language/python.rb#L136-L151).

```ruby
        # Find any Python bindings provided by recursive dependencies
        formula_deps = formula.recursive_dependencies
        pth_contents = formula_deps.map do |d|
          next if d.build? || d.test?
          # Do not add the main site-package provided by the brewed
          # Python formula, to keep the virtual-env's site-package pristine
          next if python_names.include? d.name

          dep_site_packages = Formula[d.name].opt_prefix/Language::Python.site_packages(python)
          next unless dep_site_packages.exist?

          "import site; site.addsitedir('#{dep_site_packages}')\n"
        end.compact
        unless pth_contents.empty?
          (venv_root/Language::Python.site_packages(python)/"homebrew_deps.pth").write pth_contents.join
        end
```

It just needs a little adjustment.

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb:26]: @prusnak Would you be open to a setup arg similar to `--with-system-hidapi` that would allow providing an include prefix to use system libraries rather than the vendored `hidapi`  (more context in the discussion above)?

Comment (prusnak) [reply Formula/onlykey-agent.rb:26]: Isn't --with-system-hidapi exactly that option?

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb:71]: @cho-m or @carlocab I'm spinning a bit on this and have some questions:

1. I should run this code from the formula's `def install` block before `virtualenv_install_with_resources`, right?
2. How do I get the path to _virtualenv site-packages_ (mentioned above by @cho-m) where I need write the .pth?
3. What should the .pth file's name be exactly? `homebrew-onlykey-agent-dependencies.pth` )? I am basing that on https://docs.brew.sh/Python-for-Formula-Authors#dependencies-1.
4. How exactly should I get the path to libcython (to use with `site.addsitedir`) to write into the .pth file?

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb:27]: FYI: This is resolved with the monkey patch in a8419fe04d18e6d22bf537a9f9f05a4373d309c6 . I wanted to unblock this formula for now and will consider the upstream to cython-hidapi after I get this merged.

**Comment by activescott** [comment]
I think the only thing holding up merge now is [the thread about how to properly reference cython](https://github.com/Homebrew/homebrew-core/pull/104096#discussion_r931321412) and I asked some questions in the thread about next steps.

File: Formula/onlykey-agent.rb
Comment (cho-m) [inline Formula/onlykey-agent.rb:17]: You will want to add an entry in https://github.com/Homebrew/homebrew-core/blob/master/pypi_formula_mappings.json to avoid `six` being added by `brew [bump|bump-formula-pr]`.

File: Formula/onlykey-agent.rb
Comment (cho-m) [reply Formula/onlykey-agent.rb:72]: If you want an example, I did do this for `alot` to use `notmuch` Python bindings
https://github.com/Homebrew/homebrew-core/blob/bc8a1f204ae253d407782691d27707eb8a1bc306/Formula/alot.rb#L101-L104

Before or after `virtualenv_install_with_resources` depends on if it is needed while building other dependencies or not. In `alot`, it is only a runtime dependency so could be created afterward.

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb:17]: Thanks. Done in df549621a6869ba3fc89159a84f4caffbb2beb41

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb]: I have resolved this in 56e444cb9c64135197dead6c696e9bd8cb37cadc which will replace `resource "Cython"` with `depends_on "libcython"`. Thank you for your help @cho-m !

**Comment by activescott** [comment]
@carlocab and @cho-m Thank you for all of your help and guidance 🙇. Commit 56e444cb9c64135197dead6c696e9bd8cb37cadc which will replace `resource "Cython"` with `depends_on "libcython"`. I believe that is the last outstanding item to get this ready for merge. Let me know if you find anything else.

Sorry for that last lame commit message, and I assume maintainers will rebase and squash this before merging to master. Please let me know if you'd like me to do any housekeeping like that.

File: Formula/onlykey-agent.rb
Comment (carlocab) [inline Formula/onlykey-agent.rb:209]:
```suggestion
    pth_contents = <<~EOS
      import site
      site.addsitedir('#{Formula["libcython"].opt_libexec/site_packages}')
    EOS
```
In case splitting the two commands into two lines doesn't work, we can join them again with `;`.

File: pypi_formula_mappings.json
Comment (carlocab) [inline pypi_formula_mappings.json:271]: I think you also need `cython` here.

File: Formula/onlykey-agent.rb
Comment (carlocab) [inline Formula/onlykey-agent.rb:199]: Does using `pip` to install this work? If so, we may be able to do `venv.pip_install` (which is just a wrapper around `pip install`).

**Review by carlocab — COMMENTED** [review]
Thanks for your work. I'm happy with the overall shape of this. Just a few more suggestions.

Could you also squash all your commits together when you're done? We need @BrewTestBot to merge this (in order to publish bottles simultaneously), but @BrewTestBot isn't smart enough to handle squashing commits that modify non-formula files (i.e. `pypi_formula_mappings.json`).

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb]: splitting failed it. I joined them with ; in 381e56b73b61d4858a735ff501db3ea95659f6e8 and it test started passing again on my machine.

File: pypi_formula_mappings.json
Comment (activescott) [reply pypi_formula_mappings.json]: added in f5ad421c24b031e42849b3968696ec7c944a269d

File: Formula/onlykey-agent.rb
Comment (activescott) [reply Formula/onlykey-agent.rb:199]: @carlocab We need to be able to pass the `--with-system-hidapi` option to this during install. I looked at other formula usages of `.pip_install` and didn't see any examples of that. I also looked at the source for `pip_install` and didn't notice any way to pass such an option. Am I missing something?

Comment (carlocab) [reply Formula/onlykey-agent.rb:199]: You are not. I had forgotten that we can't pass extra arguments to `#pip_install`. Thanks for looking into it!

**Review by carlocab — APPROVED** [review]
Thanks! We can merge this when CI is finished.

**Comment by activescott** [comment]
One remaining thing before merge...

- [ ] squash all your commits together

**Comment by activescott** [comment]
@carlocab Thanks again for all of your help. Do I need to do anything else to merge this?

**Comment by carlocab** [comment]
Not anymore! I just forgot to trigger a merge. (GitHub UI doesn't work because we also need the bottles to be published simultaneously.)

**Review by carlocab — APPROVED** [review]
Oops, wrong flags...

**Pull request closed without merging**
